### ISSUE TITLE
Additional Mesh Functions

### ISSUE THREAD

**Issue by bradrothenberg** [nTopology/juce_icarus#4]
Does anyone want to take a stab at writing a few the following functions operating on NTManifolds:

removeDuplicateVertices(NTManifold& mesh)

&

removeUnreferencedVertices(NTManifold& Mesh)

?

For these, we probably should use a kd-tree or R-tree to check the duplicate vertices, so this runs fast...

**Comment by bradrothenberg** [comment]
Note, maybe we should look at how VCG library does these functions & copy them : 
http://vcg.isti.cnr.it/vcglib/trimesh__smooth_8cpp_source.html

**Comment by ebwork** [comment]
Sure, will do tonight or tomorrow.
Erik

**Comment by bradrothenberg** [comment]
great -- you can check out VCG lib's implamentation here : http://vcg.isti.cnr.it/vcglib/clean_8h_source.html

**Comment by bradrothenberg** [comment]
there's also this : 

http://forum.devmaster.net/t/removal-of-duplicate-vertices/11550

**Comment by ebwork** [comment]
When you say delete duplicates, you essentially mean "merge" two vertices
that are equivalent, right?

Do you want to consider two vertices as duplicates if they are identical,
or if they are within a certain small distance of each other?

Erik

**Comment by bradrothenberg** [comment]
Yeah, maybe we should specify a threshold? ?

**Comment by ebwork** [comment]
Okay, will include equivalence threshold.  (The hashing method in the blog
you sent is based on exact match of the floating point values, so I wanted
to check with you.)

**Comment by bradrothenberg** [comment]
Oh yeah, I see that -- can you round the value, than hash it? do you think hashing is the best method??

**Comment by ebwork** [comment]
Once we said we would use thresholds, I was just going to use CGAL routines
- why reimplement the wheel?

But if I was doing it from scratch then, yes, I would probably use a
(virtual) grid with grid spacing of size equal to threshold, "round" to the
nearest grid point, use a hash to find the data structure for that grid
point, then check that d.s. and the d.s.'s for the adjacent 26 cells.
Deciding on the right way to access those adjacent cells would be the
interesting part (just do a hash search?  or add some info to the grid
point d.s.)

Erik

**Comment by bradrothenberg** [comment]
that makes sense -- like cgal's kdtree or dD subdivison??

**Comment by ebwork** [comment]
What I have started on uses k-d trees, yes.

**Comment by bradrothenberg** [comment]
awesome

**Comment by ebwork** [comment]
I added a function to NTLattice:

    NTManifoldP getVertexCleanedManifold(NTManifoldP& manifold, bool
removeUnreferenced, bool mergeDuplicates, float epsMerge);

This takes a Poly lattice as input and returns a new Poly lattice, with the
option of removing unreferenced verts, and the option of merging vertices
that are at a distance less than epsMerge from each other (note that this
nearness is taken to be transitive).

This code also uses code in the new class NTCleanVert.

There is test code in NTManifoldTest.

I pushed them on branches in two repos:
    ebwork-BRAD_LM_voro_EXOSK in Tupac_Template

and
    ebwork-master in juce_icarus

I did not implement it for Tri lattices, or do an in-place removal.  I
wanted to get this working and debugged, and ask what you want done, before
pushing it in those directions.

Erik

**Comment by bradrothenberg** [comment]
Excellent! 

should this :     
```c++
NTManifoldP getVertexCleanedManifold(NTManifoldP& manifold, bool
removeUnreferenced, bool mergeDuplicates, float epsMerge);
```

be this, if it is creating a new mesh?:
```c++
    NTManifoldP getVertexCleanedManifold(const NTManifoldP& manifold, bool
removeUnreferenced, bool mergeDuplicates, float epsMerge);
```

or are you assigning new properties to the mesh in the function??

I think it's probably a good idea for us to have an overloaded version of the function for trimeshes too ??

Do you think I should wait until monday to merge up with this?

**Comment by ebwork** [comment]
Yes should have the const on first arg.  Should be fine to merge if you
want,  I have not made any changes on this branch except for this task.

I will start on tri mesh also.

Doing family stuff Sun and Mon.   In the office Tues through Fri.
